app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php

```diff
@@ -0,0 +1,861 @@
+        $message .= "To: $recipient\n";
+        if (count($msg['recipients']['cc']) > 0) {
+            $message .= 'Cc: '.implode(',', array_keys($msg['recipients']['cc']))."\n";
+        }
+        if (count($msg['recipients']['bcc']) > 0) {
+            $message .= 'Bcc: '.implode(',', array_keys($msg['recipients']['bcc']))."\n";
+        }
+        $message .= "Content-Type: multipart/mixed; boundary=\"$separator_multipart\"\n";
+        if (isset($msg['headers'])) {
+            foreach ($msg['headers'] as $key => $value) {
+                $message .= "$key: ".$value."\n";
+            }
+        }
+        $message .= "\n--$separator_multipart\n";
+
+        $message .= "Content-Type: multipart/alternative; boundary=\"$separator\"\n";
+        if (isset($msg['text']) && strlen($msg['text']) > 0) {
+            $message .= "\n--$separator\n";
+            $message .= "Content-Type: text/plain; charset=\"UTF-8\"\n";
+            $message .= "Content-Transfer-Encoding: base64\n";
+            $message .= "\n".wordwrap(base64_encode($msg['text']), 78, "\n", true)."\n";
+        }
+        $message .= "\n--$separator\n";
+        $message .= "Content-Type: text/html; charset=\"UTF-8\"\n";
+        $message .= "\n".$msg['html']."\n";
+        $message .= "\n--$separator--\n";
+
+        foreach ($msg['attachments'] as $attachment) {
+            $message .= "--$separator_multipart\n";
+            $message .= 'Content-Type: '.$attachment['type'].'; name="'.$attachment['name']."\"\n";
+            $message .= 'Content-Disposition: attachment; filename="'.$attachment['name']."\"\n";
+            $message .= "Content-Transfer-Encoding: base64\n";
+            $message .= "\n".$attachment['content']."\n";
+        }
+
+        return $message."--$separator_multipart--";
+    }
+
+    /**
+     * @return int
+     */
+    public function getMaxBatchLimit()
+    {
+        return 50;
+    }
+
+    /**
+     * Returns a "transport" string to match the URL path /mailer/{transport}/callback.
+     *
+     * @return string
+     */
+    public function getCallbackPath()
+    {
+        return 'amazon_api';
+    }
+
+    /**
+     * @param int    $toBeAdded
+     * @param string $type
+     *
+     * @return int
+     */
+    public function getBatchRecipientCount(\Swift_Message $message, $toBeAdded = 1, $type = 'to')
+    {
+        if ($message->getTo()) {
+            $getTo = count($message->getTo());
+        } else {
+            $getTo = 0;
+        }
+
+        if ($message->getCc()) {
+            $getCc = count($message->getCc());
+        } else {
+            $getCc = 0;
+        }
+
+        if ($message->getBcc()) {
+            $getBcc = count($message->getBcc());
+        } else {
+            $getBcc = 0;
+        }
+
+        return $getTo + $getCc + $getBcc + $toBeAdded;
+    }
+
+    /**
+     * Handle bounces & complaints from Amazon.
+     *
+     * @return array
+     */
+    public function processCallbackRequest(Request $request)
```

Comment:
All these webhook processing below here are almost identical to `AmazonTransport`.

I have little experience with Mautic/PHP code, but looks like it could be extracted to a common file to avoid repetition.

Reply:
Thanks Luri for that suggestion, I did that and refactored the callback methods and moved it to a separate class.